### PR TITLE
Switch to a hug-based web API with uWSGI

### PR DESCRIPTION
**Pull request by mdcfe**
This will make life much easier when we move from dumping JSON files in the server root to using a database (MySQL/SQLite).

I'm not sure what changes will need to be made on the real ChaosBot server to get Nginx to forward uWSGI properly, but I've modified the Docker `start_services.sh` script to link the new config properly.

(Note: ~~69d7aff says "incomplete" because currently, `/` serves the hug help API and you need to visit `/index.html` to show the index.~~ Fixed in 012513b)

### CODE REVIEW

File: webserver.py [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+        __log.exception("Failed to read voters!")
+
+
+@hug.get("/meritocracy", examples=["amount=5", "amount=0"])
+def get_meritocracy(amount: hug.types.number = 20):
+    try:
+        with open("/root/workspace/Chaos/server/meritocracy.json", "r") as f:
+            meritocracy = json.load(f)
+        if amount > 0:
+            meritocracy = meritocracy[:amount]
```
Comment (rudehn) [inline webserver.py]: is meritocracy defined here?

File: webserver.py [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+__log = logging.getLogger("webserver")
+
+
+@hug.get("/voters", examples=["amount=20", "amount=0"])
+def get_voters(amount: hug.types.number = 20):
+    try:
+        with open("/root/workspace/Chaos/server/voters.json", "r") as f:
+            voters = sorted(json.load(f).items(), key=lambda x: x[1])
+        if amount > 0:
+            voters = voters[:amount]
```
Comment (rudehn) [inline webserver.py]: is voters defined here?

Comment (mdcfe) [reply webserver.py]: Yes, it is. If the sorted call returns null for some reason, it will be caught.

File: webserver.py (hunk 1 above)
Comment (mdcfe) [reply webserver.py]: Same as before

File: webserver.py (hunk 2 above)
Comment (rudehn) [reply webserver.py]: I mean is voters initially scope limited to the `with` block?

**Comment by hongaar** [comment]
Have no experience with hug or uwsgi but looks incredible, so :+1:

File: webserver.py (hunk 2 above)
Comment (mdcfe) [reply webserver.py]: I've tested it and that doesn't appear to be the case.

**Comment by mdcfe** [comment]
@hongaar I don't have much experience with them either 😛

File: webserver.py (hunk 2 above)
Comment (PlasmaPower) [reply webserver.py]: @rudehn python scopes are classes and functions only.

File: webserver.py [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+import logging
+
+
+__log = logging.getLogger("webserver")
+
+
+@hug.get("/voters", examples=["amount=20", "amount=0"])
+def get_voters(amount: hug.types.number = 20):
+    try:
+        with open("/root/workspace/Chaos/server/voters.json", "r") as f:
```
Comment (PlasmaPower) [inline webserver.py]: Cache this?

File: webserver.py [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+            voters = sorted(json.load(f).items(), key=lambda x: x[1])
+        if amount > 0:
+            voters = voters[:amount]
+        return {x[0]: x[1] for x in voters}
+    except:
+        __log.exception("Failed to read voters!")
+
+
+@hug.get("/meritocracy", examples=["amount=5", "amount=0"])
+def get_meritocracy(amount: hug.types.number = 20):
```
Comment (PlasmaPower) [inline webserver.py:30]: Do we really need a maximum amount for this?

**Comment by Swizz** [comment]
@hongaar If you take a look at the gitter chat, we choose hug and moon for the web api/site for ease.
We heavily decided to choose simplicity and minimalism first.

@md678685 Do you need me on some stuff ?

**Comment by mdcfe** [comment]
@Swizz If you want to add anything else (or you're particularly experienced with Nginx 😉) then that'd be great!

**Comment by Swizz** [comment]
Front end dev, know nothing about Nginx

**Comment by mdcfe** [comment]
Fair enough, I'll try and sort the index issue out

**Comment by mdcfe** [comment]
I think this is ready for merging - could we get a meritocracy review? @PlasmaPower

**Review by PlasmaPower — CHANGES_REQUESTED** [review]
This will currently break the top voters card, as you're requesting the first 20 voters, not the top 20 voters.

**Comment by mdcfe** [comment]
@PlasmaPower The voters are now sorted in descending (highest to lowest) order before it gets limited by the amount.

**Review by PlasmaPower — APPROVED** [review]

**Pull request merged**